File: backend/schemas/llm_observability.py
"""
Pydantic v2 response models for AI-gateway observability (Tier 1).

Every model carries the degradation envelope so a panel can render an
"unavailable" state without the UI crashing:
  - ``available`` / ``reason`` — false + reason when Loki can't be reached.
  - ``endpoint`` — the in-cluster Loki service the query was proxied to.
  - ``updated_at`` — ISO-8601 timestamp of the response.
  - ``errors`` — per-query error map for partial failures (never raised).

Data fields default to empty/zero so an unavailable envelope validates
without supplying them. ``reason`` is declared here (not in the design's
contract table) because ``response_model`` silently drops undeclared fields.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class _Envelope(BaseModel):
    """Shared degradation fields on every observability response."""

    available: bool
    endpoint: str | None = None
    updated_at: str
    reason: str | None = None
    errors: dict[str, str] = Field(default_factory=dict)


class LlmStatsResponse(_Envelope):
    total_requests: int = 0
    success_rate: float = 0.0
    avg_latency_ms: float = 0.0
    total_tokens: int = 0
    total_cost: float = 0.0
    models: int = 0


class HistogramPoint(BaseModel):
    ts: str
    value: float


class HistogramSeries(BaseModel):
    name: str
    points: list[HistogramPoint] = Field(default_factory=list)


class LlmHistogramResponse(_Envelope):
    metric: str = ""
    step_s: int = 0
    series: list[HistogramSeries] = Field(default_factory=list)


class RankingTrend(BaseModel):
    requests: float | None = None
    tokens: float | None = None
    cost: float | None = None


class RankingRow(BaseModel):
    model: str
    provider: str
    requests: int = 0
    success_rate: float = 0.0
    tokens: int = 0
    cost: float = 0.0
    avg_latency_ms: float = 0.0
    trend: RankingTrend = Field(default_factory=RankingTrend)


class LlmRankingsResponse(_Envelope):
    rows: list[RankingRow] = Field(default_factory=list)


class LlmProviderUsageResponse(_Envelope):
    """Same envelope as histogram; series keyed by inferred provider."""

    metric: str = ""
    step_s: int = 0
    series: list[HistogramSeries] = Field(default_factory=list)


class LogRow(BaseModel):
    ts: str
    type: str
    message: str
    model: str
    latency_ms: float
    prompt_tk: int
    comp_tk: int
    total_tk: int
    cost: float
    status: str
    req_body: str
    resp_body: str


class LlmLogsResponse(_Envelope):
    rows: list[LogRow] = Field(default_factory=list)
    next_end: str | None = None


class LlmFilterDataResponse(_Envelope):
    models: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)

File: frontend-v2/src/types/llm-observability.ts
/**
 * AI Gateway Observability (Tier 1: request analytics).
 *
 * Local TS types matching the backend response-shape contract in
 * `docs/superpowers/specs/…-ai-gateway-observability-design.md` §"Response shapes".
 * These are hand-written; they will be reconciled to `api-generated.ts`
 * after the backend routes land and `make openapi-types` runs.
 *
 * Every response carries the degradation envelope (`available`, `endpoint`,
 * `updated_at`, `errors`) so panels can render an unavailable state instead
 * of crashing when Loki is unreachable.
 */

/** Time-range presets mapped to Loki start/end nanosecond bounds server-side. */
export type LlmTimeRange = '1h' | '6h' | '24h' | '7d';

/** Shared degradation envelope on every observability response. */
export interface LlmEnvelope {
  available: boolean;
  endpoint: string | null;
  updated_at: string;
  errors: Record<string, string>;
}

export interface LlmStats extends LlmEnvelope {
  total_requests: number;
  success_rate: number;
  avg_latency_ms: number;
  total_tokens: number;
  total_cost: number;
  models: number;
}

export interface LlmSeriesPoint {
  ts: string;
  value: number;
}

export interface LlmSeries {
  name: string;
  points: LlmSeriesPoint[];
}

/** Metrics selectable on the histogram endpoint. */
export type LlmHistogramMetric = 'requests' | 'tokens' | 'cost' | 'models' | 'latency';

export interface LlmHistogram extends LlmEnvelope {
  metric: string;
  step_s: number;
  series: LlmSeries[];
}

export interface LlmRankingTrend {
  requests: number | null;
  tokens: number | null;
  cost: number | null;
}

export interface LlmRankingRow {
  model: string;
  provider: string;
  requests: number;
  success_rate: number;
  tokens: number;
  cost: number;
  avg_latency_ms: number;
  trend: LlmRankingTrend;
}

export interface LlmRankings extends LlmEnvelope {
  rows: LlmRankingRow[];
}

/** Provider-usage shares the histogram envelope; series are keyed by provider. */
export type LlmProviderMetric = 'cost' | 'tokens' | 'latency';
export type LlmProviderUsage = LlmHistogram;

export interface LlmLogRow {
  ts: string;
  type: string;
  message: string;
  model: string;
  latency_ms: number;
  prompt_tk: number;
  comp_tk: number;
  total_tk: number;
  cost: number;
  status: string;
  req_body: string;
  resp_body: string;
}

export interface LlmLogs extends LlmEnvelope {
  rows: LlmLogRow[];
  next_end: string | null;
}

export interface LlmFilterData extends LlmEnvelope {
  models: string[];
  statuses: string[];
}

/** Shared query params accepted by every dashboard endpoint. */
export interface LlmObservabilityParams {
  range?: LlmTimeRange;
  model?: string;
  status?: string;
}
